src/github.com/matrix-org/dendrite/clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,177 @@
+	}
+
+	dataByte, _ := json.Marshal(data)
+	err = json.Unmarshal(dataByte, &mtag)
+
+	if err != nil {
+		httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: mtag,
+	}
+}
+
+// PutTag implements PUT /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+func PutTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	userID string,
+	roomID string,
+	tag string,
+) util.JSONResponse {
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+	mtag := newMTag()
+	var properties common.TagProperties
+
+	if reqErr := httputil.UnmarshalJSONRequest(req, &properties); reqErr != nil {
+		return *reqErr
+	}
+
+	if len(data) > 0 {
+		dataByte, _ := json.Marshal(data)
+		if err := json.Unmarshal(dataByte, &mtag); err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	}
+	mtag.Tags[tag] = properties
+	addDataToDB(req, localpart, roomID, accountDB, mtag)
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// DeleteTag implements DELETE /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+func DeleteTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	userID string,
+	roomID string,
+	tag string,
+) util.JSONResponse {
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+	mtag := newMTag()
+
+	if len(data) > 0 {
+		dataByte, _ := json.Marshal(data)
+		if err := json.Unmarshal(dataByte, &mtag); err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	} else {
+		//Error indicating there is no Tag data
+		return util.JSONResponse{
+			Code: http.StatusMethodNotAllowed,
+			JSON: jsonerror.NotFound("No Tag Exists"),
+		}
+	}
+
+	// Check whether the Tag to be deleted exists
+	if _, ok := mtag.Tags[tag]; ok {
+		delete(mtag.Tags, tag)
+	} else {
+		return util.JSONResponse{
+			Code: http.StatusMethodNotAllowed,
+			JSON: struct{}{},
+		}
+	}
+	addDataToDB(req, localpart, roomID, accountDB, mtag)
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// obtainSavedTags is a utility function to get all the tags saved in the DB
+func obtainSavedTags(req *http.Request, userID string, roomID string, accountDB *accounts.Database) (string, []gomatrixserverlib.ClientEvent, error) {
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return "", []gomatrixserverlib.ClientEvent{}, err
+	}
+
+	data, err := accountDB.GetAccountDataByType(
+		req.Context(), localpart, roomID, "m.tag",
+	)
+	if err != nil {
+		return "", []gomatrixserverlib.ClientEvent{}, err
+	}
+
+	return localpart, data, nil
+}
+
+// addDataToDB is a utility function to save the tag data into the DB
+func addDataToDB(req *http.Request, localpart string, roomID string, accountDB *accounts.Database, mtag common.MTag) {
+	newTagData, _ := json.Marshal(mtag)
```

Comment:
I'd recommend checking this error and returning the error if Marshaling fails. Always better to be on the safe side.